### ISSUE TITLE
Question about the use of CMUDict in CTC finetuning

### ISSUE THREAD

**Issue by PussyCat0700** [facebookresearch/av_hubert#72]
Hi:

I am trying to reproduce finetuning using CTC loss, and I am quite uncertain how lexicon is constructed using CMUDict, could you please briefly talk about that?

In my opinion, for CTC finetuning, a linear projection layer ensuing the pretrained encoder will generate predictions of the phonemes, and the phonemes will be further translated into lexicons in some way using CMUDict, is that true?

Besides, I noticed that in the paper, the vocabulary size in CTC is 46, what does this number stand for?

**Comment by chevalierNoir** [comment]
Hi,

Yes the final projection layer maps feature into probability vector over phoneme units. We have 46 different phonemes in total, thus the vocabulary size is 46. The decoding process is same as the way used in wav2vec2 and HuBERT. Briefly, we construct a decoding graph based on lexicon and n-gram word-level language model and then run beam search on the graph. You can find a running example in the repo of [wav2vec2](https://github.com/facebookresearch/fairseq/blob/main/examples/wav2vec/README.md#evaluating-a-ctc-model) or [HuBERT](https://github.com/facebookresearch/fairseq/tree/main/examples/hubert#kenlm--fairseq-lm-decoding).

**Comment by PussyCat0700** [comment]
Hi,

Thank you for your reply!

I have run HuBERT pretraining but I found out that HuBERT is actually pretrained on letter level(e.g. dict.ltr.txt containing all the letters is required for the task to provide a dictionary of 32, while ctc loss is computed based on this dictionary and samples are taken from {train/valid}.ltr). As they have provided their word-level dictionary online for downloading [here](https://dl.fbaipublicfiles.com/fairseq/wav2vec/dict.ltr.txt), I had no problem finetuning HuBERT model.

However, it seemed that it was **not** CMUDict they were using for letter-level dictionary, and I wonder if I should just take their dictionary for finetuning, or should I construct a dictionary in some way with CMUDict on my own?

So in general I think I will just need to generate/find three additional files, **dict.ltr.txt, {train/valid}.ltr** for the finetuning to run, is that true?

**Comment by chevalierNoir** [comment]
Yes. Directly using letter dictionary is also OK. In our experiments, we find using phone units achieves slightly better performance than using letters. 

If you want to use phone, then you need to use CMUDict to map all words into its phone sequence and in turn convert the original sentence into the phone sequence, which are `{train/valid}.ltr`. `dict.ltr.txt` is of same format as [the original letter dictionary](https://dl.fbaipublicfiles.com/fairseq/wav2vec/dict.ltr.txt) but with phones.
